bxparks/bigquery-schema-generator
# Add support for existing BigQuery Schema (replace #40)

This takes the code proposed in PR #40 and attempts to address all outstanding items of concern as they were reviewed by @bxparks. #40 is very stale at this time and it does not seem that will get implemented at any point. I would like to request we close at #40 in favor of this PR but we should use that as a base of what changes I made since you last reviewed that PR.

The changes I have made as requested are:
1. Extend testdata structure to enable interpreting the existing schema as a string
    commit id: 8fbe52d0210867ec319c4a47fb9c9cd0d1ef1026
1. Convert added pytest tests to normal unittest (replaced fixtures with nested for loops which is worse but satisfies not using a different test suite)
    commit id: 1e286f14269066ce55628e143790a05f0a5a7b38
1. Delete pytest dep, and delete tox
    commit id: 1e286f14269066ce55628e143790a05f0a5a7b38
1. Add --existing_schema command line flag for reading in existing schema from a file. (super easy)
    commit id: 7afdf419a2c4a61c3b356d1b59ad5baaed8cc171
1. Write some tests that include an existing schema in testdata.txt
    commit id: fff6f5bf61a69edf67a1d19106dc5f42ef71cab4
1. Remove ERRORS INFORMED from the data_reader. This didn't make sense to me and I did not see the value. All errors should be part of the `error_logs` that are returned.
    commit id: fff6f5bf61a69edf67a1d19106dc5f42ef71cab4
1. Remove `type_mismatch_callback`. This was not properly tested and I didn't feel like maintaining/testing it.
    commit id: c5a57de54515e01de649f2705000a358f9688355
1. Fix bug as it related to bigquery not providing a mode for some nested JSON fields by defaulting to NULLABLE
    commit id: 96ca4ae148f6aa899665222d1c0e5a96968ecaae
1. Not requested but while updating the README I added the bash syntax flag to the `generate-schema` command examples


*Summarizing original changes from @bozzzzo.*
1. Break sanitize_names into it's own function and call this earlier in the schema generation process, not just during flattening.
1. Add ability to begin from an existing bigquery table schema by being able to reconstitute a schema_map from an existing bigquery schema.
